--- tracking_resource_adaptor.hpp.org	2026-02-23 19:38:45.088401318 -0500
+++ tracking_resource_adaptor.hpp	2026-02-23 19:38:24.505094659 -0500
@@ -215,7 +215,6 @@
    */
   void do_deallocate(void* ptr, std::size_t bytes, cuda_stream_view stream) noexcept override
   {
-    get_upstream_resource().deallocate(stream, ptr, bytes);
     {
       write_lock_t lock(mtx_);

@@ -246,6 +245,7 @@
         }
       }
     }
+    get_upstream_resource().deallocate(stream, ptr, bytes);
     allocated_bytes_ -= bytes;
   }
